### PR TITLE
feat: State machine

### PR DESCRIPTION
**Pull request by dnys1**
Adds the base state machine types and the skeleton for the Auth plugin's state machine

commit-id:a2731ef0

---

**Stack**:
- #1648
- #1647
- #1645
- #1642
- #1639
- #1625 ⬅

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

### CODE REVIEW

**Review by Jordan-Nelson — APPROVED** [review]

File: packages/auth/amplify_auth_cognito/lib/src/state/event/hosted_ui_event.dart [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+    CognitoSignInWithWebUIOptions options,
+    AuthProvider? provider,
+    Uri? redirectUri,
+  }) = HostedUiSignIn;
+
+  /// {@macro amplify_auth_cognito_dart.hosted_ui_cancel_sign_in}
+  const factory HostedUiEvent.cancelSignIn() = HostedUiCancelSignIn;
+
+  /// {@macro amplify_auth_cognito_dart.hosted_ui_exchange}
+  const factory HostedUiEvent.exchange(OAuthParameters paramters) =
```
Comment (ragingsquirrel3) [inline packages/auth/amplify_auth_cognito/lib/src/state/event/hosted_ui_event.dart:69]: nit: typo `paramters`

File: packages/auth/amplify_auth_cognito/lib/src/state/event/credential_store_event.dart [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+
+  @override
+  CredentialStoreEventType get type =>
+      CredentialStoreEventType.clearCredentials;
+
+  @override
+  List<Object?> get props => [type, keys];
+
+  @override
+  String? checkPrecondition(CredentialStoreState currentState) {
```
Comment (ragingsquirrel3) [inline packages/auth/amplify_auth_cognito/lib/src/state/event/credential_store_event.dart:210]: nit: this looks identical to same method in `CredentialStoreStoreCredentials`. Could it be DRYed either by moving to helper function? Or maybe a shared subclass?

File: packages/amplify_core/lib/src/state_machine/token.dart [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:meta/meta.dart';
+
+/// {@template amplify_common.token}
```
Comment (ragingsquirrel3) [inline packages/amplify_core/lib/src/state_machine/token.dart:17]: same question here and rest of file regarding amplify_common/core

File: packages/amplify_core/lib/src/state_machine/state.dart [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:aws_common/aws_common.dart';
+import 'package:meta/meta.dart';
+
+/// {@template amplify_common.state}
```
Comment (ragingsquirrel3) [inline packages/amplify_core/lib/src/state_machine/state.dart:18]: here too

**Review by ragingsquirrel3 — APPROVED** [review]
thanks, mostly looks good, approved w non-blocking nits:
* common/core in macro defs
* a typo and a nit

File: packages/auth/amplify_auth_cognito/lib/src/state/event/credential_store_event.dart (hunk 2 above)
Comment (dnys1) [reply packages/auth/amplify_auth_cognito/lib/src/state/event/credential_store_event.dart:210]: Yeah, it probably could be. Eventually, all of these event/state types will be auto-generated by Preamp.

**Pull request merged**